### DISCUSSION TITLE
how to get the exact thrust f_u

### DISCUSSION THREAD

**Discussion by LChen-99** [aerorobotics/neural-fly#4] [Q&A] [answered]
We would like to know how you get the exact thrust f_u.

How do you get the exact thrust f_u?

We also use PX4. 
![image](https://user-images.githubusercontent.com/54523369/231396513-812f0726-ca51-48a3-91db-636b4f6ac8ee.png)

**Comment by AndersonRayner** [comment]
The thrust f_u is measured using a thrust stand to get an estimate of the installed thrust for each motor.  The difference (or residual) between the observed thrust (i.e., what you have as the 'exact thrust') and the estimate is then learnt via the neural network.  In our particular implementation (due to not being able to directly measure the RPM). we just use the thrust command, which is a throttle value between 0 and 1. Then we re-scale it by (m*g/<throttle at hover>) to convert to Newtons.

**Reply by LChen-99** [reply, marked as answer]
As you said, you sending target thrust and quaternion commands to the PX4 in OFFBOARD_MODE. I can send the above command to px4 via mavlink, but I found that the mapping of throttle value to Newtons is affected by the battery. I would like to know how you solve this problem?

**Comment by LChen-99** [comment]
> The thrust f_u is measured using a thrust stand to get an estimate of the installed thrust for each motor. The difference (or residual) between the observed thrust (i.e., what you have as the 'exact thrust') and the estimate is then learnt via the neural network. In our particular implementation (due to not being able to directly measure the RPM). we just use the thrust command, which is a throttle value between 0 and 1. Then we re-scale it by (m*g/) to convert to Newtons.

Thanks for your reply.
And In your paper it is mention that you compute the acceleration by fifth-order numerical differentiation, why you don't use the acceleration values of imu.

**Comment by AndersonRayner** [comment]
We don't receive the full rate IMU data, so numerical differentiation of the EKF data was better for our system. It may have been better to use the IMU data if we had full rate and could low pass filter it.  Also, for adaptive control we used one step backwards Euler difference method. The 5 step central difference was only for training.  Either way, these are more implementation details and not necessarily the best way for every system.
